### ISSUE TITLE
Cannot upgrade UserPoolClientLambda from Nodejs14.x via the Amplify CI pipeline, therefore cannot deploy any code

### ISSUE THREAD

**Issue by dragosiordachioaia** [aws-amplify/amplify-cli#13644]
### How did you install the Amplify CLI?

npm

### If applicable, what version of Node.js are you using?

20

### Amplify CLI Version

12.10.0

### What operating system are you using?

Mac

### Did you make any manual changes to the cloud resources managed by Amplify? Please describe the changes made.

No manual changes that should be relevant to the problem I am describing

### Describe the bug

As of March 6th 2024, AWS seems not to allow Lambda functions to be created or updated if they are running on Node.js 14.x or lower. I've updated all the Lambda functions in my project to Node v20, including the one mention I was able to find of UserPoolClientLambda, in the YAML CloudFormation file for the Cognito User Pool. If I push to Amplify from my computer, the everything succeeds, all my Lambda functions show up in the Lambda console as running Node v20. If however I run the CI pipeline, it fails with this message:

> The following resources failed to deploy:
>                                  Resource Name: UserPoolClientLambda (AWS::Lambda::Function)
>                                  Event Type: create
>                                  Reason: Resource handler returned message: "The runtime parameter of nodejs14.x is no longer supported for creating or updating AWS Lambda functions. We recommend you use the new runtime (nodejs20.x) while creating or updating functions. (Service: Lambda, Status Code: 400, Request ID: 4555eef4-63e6-4439-844c-84f091c5b402)" (RequestToken: 86bafbbb-1141-6401-6589-c5df4b2b63a2, HandlerErrorCode: InvalidRequest)

I don't know how to update that function. What's worse, I cannot even find that function in the Lambda console. This project was created in 2020, so it was created with a much older version of the Amplify CLI. Deleting resources or environments and recreating them is not an option, because this application is being used in production right now and has a lot of real user data stored.

### Expected behavior

The build should succeed, or at the very least clear instructions should be provided on how to fix the problem

### Reproduction steps

I cannot provide reproduction steps, because a new Amplify project created today would not have auto-generated Lambda functions that run on Node v14.

### Project Identifier

8344d1fa6b3718c1a27d61c3551a1d8a

### Log output

<details>

```
🛑 The following resources failed to deploy:
                                 Resource Name: UserPoolClientLambda (AWS::Lambda::Function)
                                 Event Type: create
                                 Reason: Resource handler returned message: "The runtime parameter of nodejs14.x is no longer supported for creating or updating AWS Lambda functions. We recommend you use the new runtime (nodejs20.x) while creating or updating functions. (Service: Lambda, Status Code: 400, Request ID: 4555eef4-63e6-4439-844c-84f091c5b402)" (RequestToken: 86bafbbb-1141-6401-6589-c5df4b2b63a2, HandlerErrorCode: InvalidRequest)
2024-03-07T20:07:56.605Z [INFO]: 
2024-03-07T20:07:56.610Z [WARNING]: ✖ There was an error initializing your environment.
2024-03-07T20:07:56.610Z [INFO]: 🛑 Resource is not in the state stackUpdateComplete
                                 Learn more at: https://docs.amplify.aws/cli/project/troubleshooting/
2024-03-07T20:07:56.616Z [INFO]: 
2024-03-07T20:07:56.616Z [INFO]: Session Identifier: 26ba860a-eac4-499f-a21a-cfe2e58d6653


```

</details>

### Before submitting, please confirm:

- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.
- [X] I have removed any sensitive information from my code snippets and submission.

**Comment by ykethan** [comment]
Hey @dragosiordachioaia, thank you for reaching out. Does the CI pipeline also use the latest/updated Amplify CLI version?
As the `UserPoolClientLambda` is a inline function present in the auth CloudFormation templates, when running `amplify push` Amplify builds and updates the NodeJs version on the Lambda function. 

> YAML CloudFormation

It appears this is an un-migrated auth resource, you can update the resource by running `amplify update auth` which should prompt you to migrate the the auth resource.

**Comment by ykethan** [comment]
Hey @dragosiordachioaia, thank you for confirming. Could you download and un-zip the `#current-cloud-backend.zip` from the AWS S3 deployment bucket.  In the CloudFormation templates under `auth/<name>/build` and the `awscloudformation/build` do you observe a userpoolclient lambda with a NodeJs runtime as 14?

Could you try pulling the project into a new directory outside your existing project directory. In the CloudFormation templates under `backend/auth/<name>/build` and the `backend/awscloudformation/build` do you observe a userpoolclient lambda?

**Comment by dragosiordachioaia** [comment]
@ykethan I've checked all of that, no Lambda uses Node 14, only 18 and 20.

**Comment by dragosiordachioaia** [comment]
app ID is d18apgfog4c8xo. One example of a failed build (before I disabled the back-end build altogether just to get the front-end to deploy) is 1740 (I hope this is the right identifier, it seems to be an auto-incrementing build number). As for switching to `Amazon Linux: 2`, I tried it just now (while using Node 16 for the build, since building the front-end with Node 18 seems to run out of memory no matter what I try, but that's a problem to solve later) and this is the error I get:

```path /root/.nvm/versions/node/v16.20.2/lib/node_modules/aws-sdk
                                    npm ERR! command failed
                                    npm ERR! command sh -c -- node scripts/warn-maintenance-mode.js
                                    npm ERR! sh: node: command not found
                                    npm ERR! A complete log of this run can be found in:
                                    npm ERR!     /root/.npm/_logs/2024-03-11T20_41_20_388Z-debug-0.log
2024-03-11T20:41:57.270Z [ERROR]: !!! Build failed
2024-03-11T20:41:57.271Z [ERROR]: !!! Error: Command failed with exit code 127
2024-03-11T20:41:57.271Z [INFO]: # Starting environment caching...
2024-03-11T20:41:57.271Z [INFO]: # Uploading environment cache artifact...
2024-03-11T20:41:57.340Z [INFO]: # Uploaded environment cache artifact
2024-03-11T20:41:57.340Z [INFO]: # Environment caching completed
2024-03-11T20:41:57.341Z [INFO]: # A new default build image based on Amazon Linux 2023 is now generally available. The image includes newer versions of Node.js, Python, and Ruby that you can use to build your app. To use the new image, select "Amazon Linux:2023" in the "Build settings" section for your app in the Amplify console.
Terminating logging...```

**Comment by dragosiordachioaia** [comment]
As it turns out, I had a command in my backend phase to install version
10.5.0 of the Amplify CLI. This was due to an issue I had encountered in
the past for which the only fix was to pin that version of the CLI. It
seems that issue is no longer present in the latest version, so I removed
that line and switched back to the Amazon Linux 2023 build image, now
everything works again. Sorry for the confusion!
